Update docs (#43)
phenology/cgc: docs/triclustering.rst

Tri-clustering
==============

Introduction
------------

The ``triclustering`` module provides a generalization of the co-clustering algorithm to three-dimensional arrays (see

> Ref. [#]_). For geospatial data, tri-clustering analyses allow to extend the search for similarity patterns in

Reviewer: ```suggestion
Ref. [#]_). For geospatial data, tri-clustering analyses allow extending the search for similarity patterns in
```

        max_iterations=100,  # maximum number of iterations
        conv_threshold=1.e-5,  # error convergence threshold 
        nruns=10,  # number of differently-initialized runs
        epsilon=1.e-8  # numerical parameter
        output_filename='results.json'  # JSON file where to write output
    )

The input arguments of ``Triclustering`` are identical to the ``Coclustering`` ones (see :doc:`coclustering`) -
``nclusters_bnd`` is the only additional argument, which sets the maximum number of clusters along the 'band' dimension.

.. NOTE::
    The first axis of ``Z`` is assumed to represent the 'band' dimension.

Tri-clustering Implementations
------------------------------

Local (Numpy-based)
*******************

As for the co-clustering algorithm (see :doc:`coclustering`), multiple runs of the tri-clustering algorithm can be
efficiently computed in parallel using threads. In order to run the tri-clustering analysis using 4 threads:

.. code-block:: python

    results = tc.run_with_threads(nthreads=4)

Distributed (Dask-based)
************************

Also for the tri-clustering, analysis on distributed systems can be carried out using Dask (see also

> :doc:`coclustering`). Once connection to a Dask cluster is setup:

Reviewer: ```suggestion
:doc:`coclustering`). Once the connection to a Dask cluster is setup:
```

.. code-block:: python

    from dask.distributed import Client

    client = Client('tcp://daskscheduler:8786')  # connect to the Dask scheduler

the tri-clustering analysis is carried out as:

.. code-block:: python

    results = tc.run_with_dask(client)

Results
-------

The ``TriclusteringResults`` object returned by ``Triclustering.run_with_threads`` and ``Triclustering.run_with_dask``

> contains the final row, column, and band cluster assigments (``results.row_clusters``, ``results.col_clusters``, and

Reviewer: ```suggestion
contains the final row, column, and band cluster assignments (``results.row_clusters``, ``results.col_clusters``, and
```
